File: Python/1068-find_more_coins.py


class Node():
    def __init__(self):
        self.value = 0
        self.is_used = False


def find_coins(coins, total_amount):
    coins = sorted(coins, reverse=True)
    # shape: (total_amount + 1, len(coins) + 1)
    M = [[Node() for _ in range(len(coins) + 1)]
         for _ in range(total_amount + 1)]
    for v in range(1, total_amount + 1):
        for j, coin in enumerate(coins, start=1):
            if coin <= v:
                alt_v = M[v-coin][j-1].value + coin
                if alt_v >= M[v][j-1].value:
                    M[v][j].value = alt_v
                    M[v][j].is_used = True
                else:
                    M[v][j].value = M[v][j-1].value
                    M[v][j].is_used = False
            else:
                M[v][j].value = M[v][j-1].value
                M[v][j].is_used = False
    path = []
    if M[v][j].value == total_amount:
        while j > 0:
            if M[v][j].is_used:
                path.append(coins[j-1])  # idx start from 0
                v -= coins[j-1]
            j -= 1
    return path


if __name__ == '__main__':
    _, total_amount = [int(i) for i in input().split()]
    coins = [int(i) for i in input().split()]
    solution = find_coins(coins, total_amount)
    if solution:
        print(*solution)
    else:
        print("No Solution")

File: C++/1068-find_more_coins.cc
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>


struct Node {
    int value = 0;
    bool is_used = false;
};


std::vector<int>
find_coins(std::vector<int> coins, const int total_amount)
{
    std::sort(coins.rbegin(), coins.rend());
    // shape: (total_amount + 1, coins.size() + 1)
    std::vector<std::vector<Node>> M(total_amount + 1,
                                     std::vector<Node>(coins.size() + 1));
    for (int v = 1; v <= total_amount; ++v) {
        for (std::size_t j=1; j<=coins.size(); ++j) {
            int coin = coins[j-1];
            if (coin <= v) {
                int alt_v = M[v-coin][j-1].value + coin;
                if (alt_v >= M[v][j-1].value) {
                    M[v][j].value = alt_v;
                    M[v][j].is_used = true;
                } else {
                    M[v][j].value = M[v][j-1].value;
                    M[v][j].is_used = false;
                }
            } else {
                M[v][j].value = M[v][j-1].value;
                M[v][j].is_used = false;
            }
        }
    }
    std::vector<int> path;
    if (M[total_amount][coins.size()].value == total_amount) {
        int v = total_amount;
        for (std::size_t j=coins.size(); j>0; --j) {
            if (M[v][j].is_used) {
                path.push_back(coins[j-1]);
                v -= coins[j-1];
            }
        }
    }
    return path;
}


int main() {
    int n_coin, total_amount;
    std::cin >> n_coin >> total_amount;
    std::vector<int> coins;
    while (n_coin--) {
        int coin;
        std::cin >> coin;
        coins.push_back(coin);
    }

    std::vector<int> solution = find_coins(coins, total_amount);
    if (!solution.empty()) {
        auto iter = solution.begin();
        std::cout << *iter;
        for (++iter; iter!=solution.end(); ++iter) {
            std::cout << ' ' << *iter;
        }
    } else {
        std::cout << "No Solution";
    }
    std::cout << '\n';

    return 0;
}
